R/tis.R

```diff
@@ -0,0 +1,161 @@
+#'     for objects of class `tis`.
+#'   }
+#'  \item{`diagnostics`}{
+#'    A named list containing one vector:
+#'    * `pareto_k`: Not used in \code{tis}, all set to 0.
+#'    * `n_eff`: effective sample size estimates.
+#'  }
+#' }
+#'
+#' Objects of class `"tis"` also have the following [attributes][attributes()]:
+#' \describe{
+#'   \item{`norm_const_log`}{
+#'     Vector of precomputed values of `colLogSumExps(log_weights)` that are
+#'     used internally by the `weights` method to normalize the log weights.
+#'   }
+#'   \item{`r_eff`}{
+#'     If specified, the user's `r_eff` argument.
+#'   }
+#'   \item{`tail_len`}{
+#'     Not used for `tis`.
+#'   }
+#'   \item{`dims`}{
+#'     Integer vector of length 2 containing `S` (posterior sample size)
+#'     and `N` (number of observations).
+#'   }
+#'   \item{`method`}{
+#'     Method used for importance sampling, here `tis`.
+#'   }
+#' }
+#'
+#' @seealso
+#' \itemize{
+#' \item [psis()] for approximate LOO-CV using PSIS.
+#' \item [loo()] for approximate LOO-CV.
+#' \item [pareto-k-diagnostic] for PSIS diagnostics.
+#' }
+#'
+#' @references
+#' Ionides, Edward L. (2008). Truncated importance sampling.
+#' *Journal of Computational and Graphical Statistics* 17(2): 295--311.
+#'
+#' @examples
+#' log_ratios <- -1 * example_loglik_array()
+#' r_eff <- relative_eff(exp(-log_ratios))
+#' tis_result <- tis(log_ratios, r_eff = r_eff)
+#' str(tis_result)
+#'
+#' # extract smoothed weights
+#' lw <- weights(tis_result) # default args are log=TRUE, normalize=TRUE
+#' ulw <- weights(tis_result, normalize=FALSE) # unnormalized log-weights
+#'
+#' w <- weights(tis_result, log=FALSE) # normalized weights (not log-weights)
+#' uw <- weights(tis_result, log=FALSE, normalize = FALSE) # unnormalized weights
+#'
+#' @export
+tis <- function(log_ratios, ...) UseMethod("tis")
+
+#' @export
+#' @templateVar fn tis
+#' @template array
+#'
+tis.array <-
+  function(log_ratios, ...,
+           r_eff = NULL,
+           cores = getOption("mc.cores", 1)) {
+  importance_sampling.array(log_ratios = log_ratios, ...,
+                            r_eff = r_eff,
+                            cores = cores,
+                            method = "tis")
+  }
+
+#' @export
+#' @templateVar fn tis
+#' @template matrix
+#'
+tis.matrix <-
+  function(log_ratios,
+           ...,
+           r_eff = NULL,
+           cores = getOption("mc.cores", 1)) {
+    importance_sampling.matrix(log_ratios,
+                               ...,
+                               r_eff = r_eff,
+                               cores = cores,
+                               method = "tis")
+  }
+
+#' @export
+#' @templateVar fn tis
+#' @template vector
+#'
+tis.default <-
+  function(log_ratios, ..., r_eff = NULL) {
+    importance_sampling.default(log_ratios = log_ratios, ...,
+                                r_eff = r_eff, method = "tis")
+  }
+
+
+#' @rdname psis
+#' @export
+is.tis <- function(x) {
+  inherits(x, "tis") && is.list(x)
```

Comment:
Same question about `is.list` as above.

Reply:
This a copy from the `is.psis()` function that has that exact structure.

Reply:
Ok. Fair enough. Let's keep it that way for now as it didn't cause any trouble in the past.